### PR TITLE
SNOW-1296757: local testing fix udf sp issue

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-1296757

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

Fixed a bug that stored proc and udf should not remove imports already in the sys.path during the clean-up step.

### CODE REVIEW

File: tests/mock_unit/files/udf_file.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #
 # Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
 #
+import pytest
+
+pytestmark = pytest.mark.skip
```
Comment (sfc-gh-stan) [inline tests/mock_unit/files/udf_file.py:6]: Why skip?

File: tests/mock_unit/test_udf.py [hunk 2]
```diff
@@ -69,6 +69,7 @@ def add_fn(x: int, y: int) -> int:
 @pytest.mark.localtest
 def test_registering_sproc_with_qualified_identifier(session):
     custom_schema = "test_identifier_schema"
+    session.use_database("mock_database")
```
Comment (sfc-gh-stan) [inline tests/mock_unit/test_udf.py:72]: Why is this needed?

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 3]
```diff
@@ -214,10 +214,12 @@ def _import_file(
             file_path, self._session._conn.stage_registry, import_path
         )
 
-        if sproc_name:
-            self._sproc_level_imports[sproc_name].add(absolute_module_path)
-        else:
-            self._session_level_imports.add(absolute_module_path)
+        if absolute_module_path not in sys.path:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_stored_procedure.py:217]: I'm a bit confused by how this change is related to the tests in `tests/mock_unit/test_udf.py` but I see the issue. However this should be resolved using the same approach as `frozen_sys_module_keys`, i.e. saving a copy of `sys.path` before processing imports and restore in the exit stack.

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_stored_procedure.py:217]: https://github.com/snowflakedb/snowpark-python/blob/93b32a1ef04181c56b18e02441af6ffdb4b2dfce/src/snowflake/snowpark/mock/_stored_procedure.py#L127C9-L127C31

File: tests/mock_unit/files/udf_file.py (hunk 1 above)
Comment (sfc-gh-aling) [reply tests/mock_unit/files/udf_file.py:6]: yes, if not adding this mark, this is picked up as a test file.
or it can be ignore using --ignore option in tox file, but in my test it leads to the file missing when we try to register this file for a udf.

File: src/snowflake/snowpark/mock/_plan.py [hunk 4]
```diff
@@ -453,17 +453,21 @@ def handle_udf_expression(
 
     # Save a copy of module cache
     frozen_sys_module_keys = set(sys.modules.keys())
+    # Save a copy of sys path
+    frozen_sys_path = list(sys.path)
 
     def cleanup_imports():
+        added_path = set(sys.path) - set(frozen_sys_path)
+
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_plan.py:461]:
```suggestion
```

File: src/snowflake/snowpark/mock/_stored_procedure.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stored_procedure.py]: updated the implementation, thanks for pointing

**Review by sfc-gh-jrose — APPROVED** [review]

**Review by sfc-gh-stan — APPROVED** [review]
